### PR TITLE
Make WebRequest.GetSystemWebProxy() return a working proxy

### PR DESCRIPTION
**Pull request by stephentoub**
Today it returns a singleton on .NET Core that throws PlatformNotSupportedExceptions, and is only used for its reference identity by the WebRequest implementation.  This is problematic for existing code that's taking this proxy and passing it to, for example, ClientWebSocket's Proxy, as it will end up being an expensive nop.  In fact, there's currently no way when targeting netstandard2.0 to tell ClientWebSocket to use a default proxy.

With this fix, WebRequest.DefaultWebProxy (which just defaults to GetSystemWebProxy()) will now return HttpClient.DefaultProxy, which is actually a working proxy implementation.

This should hopefully address https://github.com/Azure/azure-sdk-for-net/issues/7899 once ported back to release/3.0.

### CODE REVIEW

File: src/System.Net.Requests/src/System.Net.Requests.csproj [hunk 1]
```diff
@@ -22,7 +22,6 @@
     <Compile Include="System\Net\IWebRequestCreate.cs" />
     <Compile Include="System\Net\ProtocolViolationException.cs" />
     <Compile Include="System\Net\RequestStream.cs" />
-    <Compile Include="System\Net\SystemWebProxy.cs" />
```
Comment (davidsh) [inline src/System.Net.Requests/src/System.Net.Requests.csproj]: Removing this class will break this code in WinHttpHandler:

https://github.com/dotnet/corefx/blob/fe8db47cb290eff654e8419ce6d615e9bb2ce189/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L709-L714

We can choose to do that. But we should adjust the related WinHttpHandler code (perhaps use HttpClient.DefaultProxy now).

Comment (stephentoub) [reply src/System.Net.Requests/src/System.Net.Requests.csproj]: I assumed UseWinInetProxy had the same behavior as HttpClient.DefaultProxy; does it not?  If it does, then it wouldn't actually break it... the proxy will simply work and we'll never end up in the catch block.

Comment (davidsh) [reply src/System.Net.Requests/src/System.Net.Requests.csproj]: We also might determine that once #41182 is implemented, then we don't need to support this behavior since callers would be using WinHttpHandler explicitly and could set `UseWinInetProxy` directly.

Comment (davidsh) [reply src/System.Net.Requests/src/System.Net.Requests.csproj]: >I assumed UseWinInetProxy had the same behavior as HttpClient.DefaultProxy (when no environment variables were defined, as they'd take precedence); does it not?

Yes, but only in the initial default case of using HttpClient.DefaultProxy. That property also has a setter. So, someone could set it to any custom IWebProxy.

Comment (davidsh) [reply src/System.Net.Requests/src/System.Net.Requests.csproj]: I'm trying to think through the various combinations here. Perhaps everything will be ok after removing this code. But we might need to revisit this change afterwards.

Comment (stephentoub) [reply src/System.Net.Requests/src/System.Net.Requests.csproj]: > So, someone could set it to any custom IWebProxy.

Sure, but that's no different from today.

So I'm not seeing how anything is negatively affected.

Comment (davidsh) [reply src/System.Net.Requests/src/System.Net.Requests.csproj]: Correct. I'm thinking in real-time with these comments we're discussing. So, I won't block on this right now. It's just something that occurred to me after looking at this PR.

Comment (stephentoub) [reply src/System.Net.Requests/src/System.Net.Requests.csproj]: > Correct. I'm thinking in real-time with these comments we're discussing

Heh, ok. :smile:

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/MacProxy.cs [hunk 2]
```diff
@@ -17,11 +17,7 @@ namespace System.Net.Http
 {
     internal sealed class MacProxy : IWebProxy
     {
-        public ICredentials Credentials
-        {
-            get => null;
-            set => throw new NotSupportedException();
-        }
+        public ICredentials Credentials { get; set; }
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/MacProxy.cs]: fyi: For the Mac, the work left in #26593 would support at least the 'get' operator here once explicit credentials are read from the Mac system settings.

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/MacProxy.cs]: Yup, thanks.

I changed this after some tests were failing that were getting/setting WebRequest.DefaultWebProxy's Credentials.  And in looking around at other IWebProxy implementations and the docs, throwing from set_Credentials is not something we do.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs [hunk 3]
```diff
@@ -283,7 +283,10 @@ public ICredentials Credentials
             {
                 return _credentials;
```
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs]: We've had some requests that we parse the username/password out of URIs that might be set into these environment variables similar to how libcurl works (for explicit credentials anyways).

Example:

```
set HTTP_PROXY=http://username:password@mylocalproxy.corp.contoso.com:8080/
```

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs]: Aren't we already doing that?
https://github.com/dotnet/corefx/blob/516f8508791f8648359841d5cc6ed9857e27c375/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L109

Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs]: Thanks for pointing that out. So, it's just the Mac we aren't reading explicit credentials from system settings.

Do you recall if we have any unit tests for this? I.e. setting environment variables and then testing that the credentials are being properly set/get into the proxy's GetCredentials() method?

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs]: https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/tests/UnitTests/HttpEnvironmentProxyTest.cs

?

Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs]: ok. thanks. Looks like we have unit tests but no end-to-end tests.

**Comment by stephentoub** [comment]
@geoffkizer, mind taking a look at this? Any concerns?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs [hunk 4]
```diff
@@ -283,7 +283,10 @@ public ICredentials Credentials
...
             }
-            set { throw new NotSupportedException(); }
+            set
+            {
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs]: Why did you add support for set here (and elsewhere)?

I'm not sure what the right behavior here is, I just want to understand the rationale.

Do we support setting this on Framework when you call GetSystemWebProxy?

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs]: > Do we support setting this on Framework when you call GetSystemWebProxy?

Yup.  `GetSystemWebProxy` just returns a `WebProxy` (well, a wrapper around one that just delegates to it)
https://referencesource.microsoft.com/#System/net/System/Net/WebRequest.cs,1092
and WebProxy.set_Credentials just stores them:
https://referencesource.microsoft.com/#System/net/System/Net/webproxy.cs,227

> Why did you add support for set here (and elsewhere)?

I initially hadn't, and then a bunch of WebRequest tests started failing because they expected to be able to.  That led me to look around at what's done in netfx and in other proxy implementations, and they pretty much all just support roundtripping the set credentials.  I didn't find any that threw NotSupportedException like these ones are doing.  And the docs don't say anything about NotSupportedException.  So, it seemed like more of a pit of failure to throw than to just let the set value roundtrip.

Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs]: Ok, thanks.
